### DISCUSSION TITLE
question: failed to find timezone 'Asia/Shanghai' in time zone database

### DISCUSSION THREAD

**Discussion by TCeason** [BurntSushi/jiff#160] [Q&A] [answered]
**Summary**

Now I think jiff will get timezone info with TimeZone::get() accross https://github.com/BurntSushi/jiff/blob/a44e494ff2ab0fa5d135fffa9d29ca5424aee1ba/src/tz/db/mod.rs#L235

```
static ZONEINFO_DIRECTORIES: &[&str] =
    &["/usr/share/zoneinfo", "/etc/zoneinfo"];
```

But actually, I get err:

```
failed to find timezone 'Asia/Shanghai' in time zone database
```

And in Chrono's Tz parse, it's ok. Looks like it generate a tzfile.

https://github.com/chronotope/chrono-tz/blob/0acd431a547f87446cd9338eb492ec8d758f59c2/chrono-tz-build/src/lib.rs#L561

And I try to println  tz::db().available(), it's empty.

**Comment by TCeason** [comment]
Maybe I missed some key configuration?

**Comment by TCeason** [comment]
This docker image not support tzdata. 


```shell
$ docker pull summerwind/actions-runner-dind:v2.320.0-ubuntu-22.04

$ docker images              
REPOSITORY                         TAG                     IMAGE ID       CREATED         SIZE
summerwind/actions-runner-dind     v2.320.0-ubuntu-22.04   4da10987fec9   4 weeks ago     1.22GB

$ docker run -it 4da10987fec9  /bin/bash
To run a command as administrator (user "root"), use "sudo <command>".
See "man sudo_root" for details.

runner@6510b467a532:/$ ls  /usr/share/zoneinfo
ls: cannot access '/usr/share/zoneinfo': No such file or directory
runner@6510b467a532:/$ 
exit

```

So that means if I want to use jiff crate, must install tzdata first ?

**Comment by TCeason** [comment]
> This docker image not support tzdata.
> 
> ```shell
> $ docker pull summerwind/actions-runner-dind:v2.320.0-ubuntu-22.04
> 
> $ docker images              
> REPOSITORY                         TAG                     IMAGE ID       CREATED         SIZE
> summerwind/actions-runner-dind     v2.320.0-ubuntu-22.04   4da10987fec9   4 weeks ago     1.22GB
> 
> $ docker run -it 4da10987fec9  /bin/bash
> To run a command as administrator (user "root"), use "sudo <command>".
> See "man sudo_root" for details.
> 
> runner@6510b467a532:/$ ls  /usr/share/zoneinfo
> ls: cannot access '/usr/share/zoneinfo': No such file or directory
> runner@6510b467a532:/$ 
> exit
> ```
> 
> So that means if I want to use jiff crate, must install tzdata first ?

Is it by design? Why we need it?

**Comment by BurntSushi** [comment, marked as answer]
As a practical matter, if you enable logging, Jiff is should emit very detailed log messages about what's going wrong when it comes to system integration stuff. That probably would have saved you some head-scratching.

> Is it by design? Why we need it?

As a conceptual matter, think about it: every time zone has its own rules for transitions. How does your code know what they are?

On Unix systems, the standard way to do this is with tzdb at `/usr/share/zoneinfo`. In your case, installing it into your Docker image should fix your problem. **This is the preferred approach.**

`chrono-tz` doesn't need this because it bundles all of the data into your Rust program. Jiff supports this too, and you can enable it with the [`tzdb-bundle-always` crate feature](https://docs.rs/jiff/latest/jiff/index.html#time-zone-features).

Otherwise, the [platform documentation](https://docs.rs/jiff/latest/jiff/_documentation/platform/index.html) goes to great pains to try and explain everything here. Your specific situation is even explicitly mentioned!

> The existence of /usr/share/zoneinfo is not guaranteed in all Unix environments. For example, stripped down Docker containers might omit a full copy of the time zone database. Jiff will still work in such environments, but all IANA time zone identifier lookups will fail. To fix this, you can either install the IANA Time Zone Database into your environment, or you can enable the Jiff crate feature tzdb-bundle-always. This compile time setting will cause Jiff to depend on jiff-tzdb, which includes a complete copy of the IANA Time Zone Database embedded into the compiled artifact.

**Reply by BurntSushi** [reply]
@TCeason In the future, please use Discussions to ask questions. I converted your issue to one.

**Reply by TCeason** [reply]
> @TCeason In the future, please use Discussions to ask questions. I converted your issue to one.

Cool I got it~

**Reply by LinusU** [reply]
@BurntSushi how does one enable logging? I'm setting up a `tracing_subscriber` in my program with the `INFO` level, but this seems not to be sufficient?

I'm currently getting `Error: failed to find timezone 'Europe/Stockholm' in time zone database` even though the `/usr/share/zoneinfo/Europe/Stockholm` file exists. This is on Ubuntu 16.04.7 LTS, which is super old, but it's what I have to work with 🙈 

edit: seems like `logging` is a feature flag, trying that now!

**Reply by BurntSushi** [reply]
Yeah you need `logging`, otherwise Jiff would need to unconditionally depend on the `log` crate.

**Reply by LinusU** [reply]
I've added the logging feature, but it itself doesn't output anything.

After, I added the crate [`tracing_log`](https://docs.rs/tracing-log/latest/tracing_log/) and the code `LogTracer::init()?;` to get it to output to my already setup tracing log.

It would be great to have a bit about this in the readme! I'm very short on time right now but will try to submit a PR unless someone beats me to it.

-----

Further, I'm getting the following error:

```text
WARN failed to cache time zone from file /usr/share/zoneinfo/Europe/Stockholm: /usr/share/zoneinfo/Europe/Stockholm: transition second value -576460752303423488 is out of range: parameter 'second' with value -576460752303423488 is not in the required range of -377705023201..=253402207200
```

@BurntSushi would it be prudent to open an issue about this, or is my system (Ubuntu 16) simply too old for you to want to deal with? 😅 

(edit: for the record, it works on macOS 15)

**Reply by BurntSushi** [reply]
The problem with docs about logging is that it varies wildly depending on which crate you're using. I think that's probably beyond the scope of what the Jiff docs can reasonably provide. Getting logging working is really up to the log crates to specify.

Otherwise, yes, please do file an issue. Please include:

* The version of your `tzdata` package.
* The `/usr/share/zoneinfo/Europe/Stockholm` file. It's a binary file, so if you could attach it or base64 encode it if GitHub doesn't like binary files.
* The full logging output at `TRACE` level.

**Reply by BurntSushi** [reply]
Someone else was able to provide a proper reproduction: https://github.com/BurntSushi/jiff/issues/163

This should be fixed in `jiff 0.1.15`.
